File: cli/internal/connectapi/room_directory.go
package connectapi

import (
	"context"

	"connectrpc.com/connect"
	"hmans.de/chatto/internal/core"
	apiv1 "hmans.de/chatto/internal/pb/chatto/api/v1"
	corev1 "hmans.de/chatto/internal/pb/chatto/core/v1"
)

type roomDirectoryService struct {
	api *API
}

func (s *roomDirectoryService) ListRooms(ctx context.Context, req *connect.Request[apiv1.ListRoomsRequest]) (*connect.Response[apiv1.ListRoomsResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}

	rooms, err := s.api.core.RoomDirectoryReads().ListRooms(ctx, caller.UserID, core.RoomDirectoryListOptions{
		IncludeChannels: roomDirectoryScopeIncludesChannels(req.Msg.GetScope()),
		IncludeDMs:      roomDirectoryScopeIncludesDMs(req.Msg.GetScope()),
	})
	if err != nil {
		return nil, connectError(err)
	}

	apiRooms := make([]*apiv1.RoomWithViewerState, 0, len(rooms))
	for _, room := range rooms {
		apiRooms = append(apiRooms, apiRoomWithViewerState(room))
	}

	return connect.NewResponse(&apiv1.ListRoomsResponse{Rooms: apiRooms}), nil
}

func (s *roomDirectoryService) ListRoomGroups(ctx context.Context, req *connect.Request[apiv1.ListRoomGroupsRequest]) (*connect.Response[apiv1.ListRoomGroupsResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}

	groups, err := s.api.core.RoomDirectoryReads().ListRoomGroups(ctx, caller.UserID, core.RoomDirectoryGroupOptions{})
	if err != nil {
		return nil, connectError(err)
	}

	apiGroups := make([]*apiv1.RoomGroup, 0, len(groups))
	for _, group := range groups {
		apiGroups = append(apiGroups, apiRoomGroup(group))
	}
	return connect.NewResponse(&apiv1.ListRoomGroupsResponse{Groups: apiGroups}), nil
}

func (s *roomDirectoryService) GetRoomGroup(ctx context.Context, req *connect.Request[apiv1.GetRoomGroupRequest]) (*connect.Response[apiv1.GetRoomGroupResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}

	group, err := s.api.core.RoomDirectoryReads().GetRoomGroup(ctx, caller.UserID, req.Msg.GetGroupId(), core.RoomDirectoryGroupOptions{})
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&apiv1.GetRoomGroupResponse{Group: apiRoomGroup(group)}), nil
}

func (s *roomDirectoryService) BatchGetRoomGroups(ctx context.Context, req *connect.Request[apiv1.BatchGetRoomGroupsRequest]) (*connect.Response[apiv1.BatchGetRoomGroupsResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}

	groups, err := s.api.core.RoomDirectoryReads().BatchGetRoomGroups(ctx, caller.UserID, req.Msg.GetGroupIds(), core.RoomDirectoryGroupOptions{})
	if err != nil {
		return nil, connectError(err)
	}

	apiGroups := make([]*apiv1.RoomGroup, 0, len(groups))
	for _, group := range groups {
		apiGroups = append(apiGroups, apiRoomGroup(group))
	}
	return connect.NewResponse(&apiv1.BatchGetRoomGroupsResponse{Groups: apiGroups}), nil
}

func (s *roomDirectoryService) GetRoom(ctx context.Context, req *connect.Request[apiv1.GetRoomRequest]) (*connect.Response[apiv1.GetRoomResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}

	room, err := s.api.core.RoomDirectoryReads().GetRoom(ctx, caller.UserID, req.Msg.GetRoomId())
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&apiv1.GetRoomResponse{Room: apiRoomWithViewerState(room)}), nil
}

func (s *roomDirectoryService) BatchGetRooms(ctx context.Context, req *connect.Request[apiv1.BatchGetRoomsRequest]) (*connect.Response[apiv1.BatchGetRoomsResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}

	rooms, err := s.api.core.RoomDirectoryReads().BatchGetRooms(ctx, caller.UserID, req.Msg.GetRoomIds())
	if err != nil {
		return nil, connectError(err)
	}

	apiRooms := make([]*apiv1.RoomWithViewerState, 0, len(rooms))
	for _, room := range rooms {
		apiRooms = append(apiRooms, apiRoomWithViewerState(room))
	}
	return connect.NewResponse(&apiv1.BatchGetRoomsResponse{Rooms: apiRooms}), nil
}

func apiRoomWithViewerState(room *core.DirectoryRoom) *apiv1.RoomWithViewerState {
	if room == nil {
		return nil
	}
	state := room.ViewerState
	return &apiv1.RoomWithViewerState{
		Room: apiRoom(room.Room),
		ViewerState: &apiv1.RoomViewerState{
			IsMember:  state.IsMember,
			HasUnread: state.HasUnread,
			Permissions: permissionGrants(
				permissionGrant(core.PermRoomList, state.CanListRoom),
				permissionGrant(core.PermRoomJoin, state.CanJoinRoom),
				permissionGrant(core.PermMessagePost, state.CanPostMessage),
				permissionGrant(core.PermMessagePostInThread, state.CanPostInThread),
				permissionGrant(core.PermMessageAttach, state.CanAttach),
				permissionGrant(core.PermMessageReact, state.CanReact),
				permissionGrant(core.PermMessageEcho, state.CanEchoMessage),
				permissionGrant(core.PermMessageManage, state.CanManageOthersMessage),
				permissionGrant(core.PermRoomManage, state.CanManageRoom),
				permissionGrant(core.PermRoomMemberBan, state.CanBanRoomMembers),
			),
		},
	}
}

func apiRoomGroup(group *core.DirectoryRoomGroup) *apiv1.RoomGroup {
	if group == nil || group.Group == nil {
		return nil
	}
	apiGroup := &apiv1.RoomGroup{
		Id:          group.Group.GetId(),
		Name:        group.Group.GetName(),
		Description: group.Group.GetDescription(),
		ViewerState: &apiv1.RoomGroupViewerState{
			Permissions: permissionGrants(
				permissionGrant(core.PermRoomCreate, group.ViewerState.CanCreateRoom),
				permissionGrant(core.PermRoomManage, group.ViewerState.CanManageRoomGroup),
			),
		},
	}
	for _, item := range group.Items {
		switch {
		case item.Room != nil:
			apiGroup.Items = append(apiGroup.Items, &apiv1.RoomGroupItem{
				Item: &apiv1.RoomGroupItem_Room{Room: apiRoomWithViewerState(item.Room)},
			})
		case item.SidebarLink != nil:
			apiGroup.Items = append(apiGroup.Items, &apiv1.RoomGroupItem{
				Item: &apiv1.RoomGroupItem_SidebarLink{SidebarLink: apiSidebarLink(item.SidebarLink)},
			})
		}
	}
	return apiGroup
}

func permissionGrant(permission core.Permission, granted bool) *apiv1.PermissionGrant {
	return &apiv1.PermissionGrant{
		Permission: string(permission),
		Granted:    granted,
	}
}

func permissionGrants(grants ...*apiv1.PermissionGrant) []*apiv1.PermissionGrant {
	return grants
}

func roomDirectoryScopeIncludesChannels(scope apiv1.RoomDirectoryScope) bool {
	return scope == apiv1.RoomDirectoryScope_ROOM_DIRECTORY_SCOPE_UNSPECIFIED ||
		scope == apiv1.RoomDirectoryScope_ROOM_DIRECTORY_SCOPE_ALL ||
		scope == apiv1.RoomDirectoryScope_ROOM_DIRECTORY_SCOPE_CHANNELS
}

func roomDirectoryScopeIncludesDMs(scope apiv1.RoomDirectoryScope) bool {
	return scope == apiv1.RoomDirectoryScope_ROOM_DIRECTORY_SCOPE_UNSPECIFIED ||
		scope == apiv1.RoomDirectoryScope_ROOM_DIRECTORY_SCOPE_ALL ||
		scope == apiv1.RoomDirectoryScope_ROOM_DIRECTORY_SCOPE_DMS
}

func apiSidebarLink(link *corev1.SidebarLink) *apiv1.SidebarLink {
	if link == nil {
		return nil
	}
	return &apiv1.SidebarLink{
		Id:    link.GetId(),
		Label: link.GetLabel(),
		Url:   link.GetUrl(),
	}
}

File: apps/frontend/src/lib/api-client/roomDirectory.ts
import {
  authHeaders,
  Code,
  ConnectError,
  createChattoClient,
  handleAuthError,
  type ConnectAPIConfig
} from './connect.js';
import { RoomDirectoryService } from '@chatto/api-types/api/v1/room_directory_connect';
import type {
  RoomGroup,
  RoomGroupItem,
  RoomGroupViewerState,
  RoomViewerState,
  RoomWithViewerState
} from '@chatto/api-types/api/v1/room_directory_pb';
import { RoomDirectoryScope } from '@chatto/api-types/api/v1/room_directory_pb';
import { RoomKind } from '@chatto/api-types/api/v1/rooms_pb';

export type RoomDirectoryAPIConfig = ConnectAPIConfig;

export type DirectoryRoomSummary = {
  id: string;
  name: string;
  description: string | null;
  kind: RoomKind;
  archived: boolean;
  isUniversal: boolean;
  isMember: boolean;
  hasUnread: boolean;
  canJoinRoom: boolean;
  canManageRoom: boolean;
};

export type DirectoryRoomDetails = DirectoryRoomSummary & {
  canPostMessage: boolean;
  canPostInThread: boolean;
  canAttach: boolean;
  canReact: boolean;
  canEchoMessage: boolean;
  canManageOthersMessage: boolean;
  canBanRoomMembers: boolean;
};

export type DirectorySidebarLink = {
  id: string;
  label: string;
  url: string;
};

export type DirectoryRoomGroupItem =
  | {
      id: string;
      type: 'room';
      roomId: string;
      room: DirectoryRoomSummary;
    }
  | {
      id: string;
      type: 'link';
      link: DirectorySidebarLink;
    };

export type DirectoryRoomGroup = {
  id: string;
  name: string;
  canCreateRoom: boolean;
  canManageGroup: boolean;
  roomIds: string[];
  items: DirectoryRoomGroupItem[];
};

export { RoomDirectoryScope };
export { RoomKind };

const RoomPermission = {
  Attach: 'message.attach',
  BanMember: 'room.ban-member',
  CreateRoom: 'room.create',
  EchoMessage: 'message.echo',
  JoinRoom: 'room.join',
  ManageMessage: 'message.manage',
  ManageRoom: 'room.manage',
  PostInThread: 'message.post-in-thread',
  PostMessage: 'message.post',
  React: 'message.react'
} as const;

export function createRoomDirectoryAPI(config: RoomDirectoryAPIConfig) {
  const directory = createChattoClient(RoomDirectoryService, config);
  const headers = () => authHeaders(config);

  return {
    async listRooms(scope: RoomDirectoryScope): Promise<DirectoryRoomSummary[]> {
      try {
        const response = await directory.listRooms({ scope }, { headers: headers() });
        return response.rooms.flatMap((entry) => mapDirectoryRoom(entry) ?? []);
      } catch (err) {
        return handleAuthError(config, err);
      }
    },

    async getRoom(roomId: string): Promise<DirectoryRoomDetails | null> {
      try {
        const response = await directory.getRoom({ roomId }, { headers: headers() });
        return mapDirectoryRoomDetails(response.room);
      } catch (err) {
        if (err instanceof ConnectError && err.code === Code.NotFound) {
          return null;
        }
        return handleAuthError(config, err);
      }
    },

    async batchGetRooms(roomIds: string[]): Promise<DirectoryRoomDetails[]> {
      try {
        const response = await directory.batchGetRooms({ roomIds }, { headers: headers() });
        return response.rooms.flatMap((entry) => {
          const mapped = mapDirectoryRoomDetails(entry);
          return mapped ? [mapped] : [];
        });
      } catch (err) {
        return handleAuthError(config, err);
      }
    },

    async listRoomGroups(): Promise<DirectoryRoomGroup[]> {
      try {
        const response = await directory.listRoomGroups({}, { headers: headers() });
        return response.groups.map(mapRoomGroup);
      } catch (err) {
        return handleAuthError(config, err);
      }
    },

    async getRoomGroup(groupId: string): Promise<DirectoryRoomGroup | null> {
      try {
        const response = await directory.getRoomGroup({ groupId }, { headers: headers() });
        return response.group ? mapRoomGroup(response.group) : null;
      } catch (err) {
        if (err instanceof ConnectError && err.code === Code.NotFound) {
          return null;
        }
        return handleAuthError(config, err);
      }
    },

    async batchGetRoomGroups(groupIds: string[]): Promise<DirectoryRoomGroup[]> {
      try {
        const response = await directory.batchGetRoomGroups({ groupIds }, { headers: headers() });
        return response.groups.map(mapRoomGroup);
      } catch (err) {
        return handleAuthError(config, err);
      }
    }
  };
}

export type RoomDirectoryAPI = ReturnType<typeof createRoomDirectoryAPI>;

export function mapDirectoryRoomDetails(
  entry: RoomWithViewerState | undefined
): DirectoryRoomDetails | null {
  if (!entry) return null;

  const summary = mapDirectoryRoom(entry);
  if (!summary) return null;

  return {
    ...summary,
    canPostMessage: hasRoomPermission(entry.viewerState, RoomPermission.PostMessage),
    canPostInThread: hasRoomPermission(entry.viewerState, RoomPermission.PostInThread),
    canAttach: hasRoomPermission(entry.viewerState, RoomPermission.Attach),
    canReact: hasRoomPermission(entry.viewerState, RoomPermission.React),
    canEchoMessage: hasRoomPermission(entry.viewerState, RoomPermission.EchoMessage),
    canManageOthersMessage: hasRoomPermission(entry.viewerState, RoomPermission.ManageMessage),
    canManageRoom: hasRoomPermission(entry.viewerState, RoomPermission.ManageRoom),
    canBanRoomMembers: hasRoomPermission(entry.viewerState, RoomPermission.BanMember)
  };
}

export function mapDirectoryRoom(entry: RoomWithViewerState): DirectoryRoomSummary | null {
  if (!entry.room) return null;
  return {
    id: entry.room.id,
    name: entry.room.name,
    description: entry.room.description || null,
    kind: entry.room.kind,
    archived: entry.room.archived,
    isUniversal: entry.room.universal,
    isMember: entry.viewerState?.isMember ?? false,
    hasUnread: entry.viewerState?.hasUnread ?? false,
    canJoinRoom: hasRoomPermission(entry.viewerState, RoomPermission.JoinRoom),
    canManageRoom: hasRoomPermission(entry.viewerState, RoomPermission.ManageRoom)
  };
}

export function mapRoomGroup(group: RoomGroup): DirectoryRoomGroup {
  return {
    id: group.id,
    name: group.name,
    canCreateRoom: hasRoomGroupPermission(group.viewerState, RoomPermission.CreateRoom),
    canManageGroup: hasRoomGroupPermission(group.viewerState, RoomPermission.ManageRoom),
    roomIds: uniqueRoomIds(group.items),
    items: sidebarItemsFromAPI(group)
  };
}

function hasRoomPermission(state: RoomViewerState | undefined, permission: string): boolean {
  return (
    state?.permissions.some((grant) => grant.permission === permission && grant.granted) ?? false
  );
}

function hasRoomGroupPermission(
  state: RoomGroupViewerState | undefined,
  permission: string
): boolean {
  return (
    state?.permissions.some((grant) => grant.permission === permission && grant.granted) ?? false
  );
}

function uniqueRoomIds(items: readonly RoomGroupItem[]): string[] {
  const seen: Record<string, true> = Object.create(null);
  return items.flatMap((item) => {
    if (item.item.case !== 'room') return [];
    const id = item.item.value.room?.id;
    if (!id || seen[id]) return [];
    seen[id] = true;
    return [id];
  });
}

function sidebarItemsFromAPI(group: RoomGroup): DirectoryRoomGroupItem[] {
  return group.items.flatMap((item) => mapRoomGroupItem(item) ?? []);
}

function mapRoomGroupItem(item: RoomGroupItem): DirectoryRoomGroupItem | null {
  if (item.item.case === 'room') {
    const roomId = item.item.value.room?.id;
    const room = mapDirectoryRoom(item.item.value);
    return roomId && room ? { id: `room:${roomId}`, type: 'room', roomId, room } : null;
  }
  if (item.item.case === 'sidebarLink') {
    return {
      id: `link:${item.item.value.id}`,
      type: 'link',
      link: {
        id: item.item.value.id,
        label: item.item.value.label,
        url: item.item.value.url
      }
    };
  }
  return null;
}
